src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs

```diff
@@ -118,13 +118,33 @@ public string ClientVersion
 
         public bool RetryOnServerErrors { get; set; } = true;
 
-        public ManagedIdentityId ManagedIdentityId { get; internal set; }
+        public Func<AppTokenProviderParameters, Task<AppTokenProviderResult>> AppTokenProvider;
 
+        #region ManagedIdentity
+        public ManagedIdentityId ManagedIdentityId { get; internal set; }
         public bool IsManagedIdentity { get; }
+#if TRA
+        public Credential.KeyMaterialInfo KeyMaterialInfo { get; internal set; }

```

Comment:
I don't think this is the right place for this. ApplicationConfiguration is for whatever the user configures. KeyMaterialInfo is not something we expose. This is only used in `CredentialSource.cs` (for READ) and in the builder (for WRITE) so I would argue it should all go to `CredentialSource`.

Single Responsability Principle.

Reply:
removed but will discuss if the new approach (PlatformProxy) is the correct approach